src/stan/mcmc/chains.hpp

```diff
@@ -254,22 +256,39 @@ namespace stan {
           var_plus += variance(chain_mean);
         Eigen::VectorXd rho_hat_t(n_samples);
         rho_hat_t.setZero();
-        double rho_hat = 0;
-        int max_t = 0;
-        for (int t = 1; (t < n_samples && rho_hat >= 0); t++) {
-          Eigen::VectorXd acov_t(chains);
-          for (int chain = 0; chain < chains; chain++) {
-            acov_t(chain) = acov(chain)(t);
+        Eigen::VectorXd acov_t(chains);
+        for (int chain = 0; chain < chains; chain++)
+          acov_t(chain) = acov(chain)(1);
+        double rho_hat_even = 1;
+        double rho_hat_odd = 1 - (mean_var - mean(acov_t)) / var_plus;
+        rho_hat_t(1) = rho_hat_odd;
+        // Geyer's initial positive sequence
+        int max_t = 1;
+        for (int t = 1;
+             (t < (n_samples - 2) && (rho_hat_even + rho_hat_odd) >= 0);
+             t += 2) {
+          for (int chain = 0; chain < chains; chain++)
+            acov_t(chain) = acov(chain)(t + 1);
+          rho_hat_even = 1 - (mean_var - mean(acov_t)) / var_plus;
+          for (int chain = 0; chain < chains; chain++)
+            acov_t(chain) = acov(chain)(t + 2);
+          rho_hat_odd = 1 - (mean_var - mean(acov_t)) / var_plus;
+          if ((rho_hat_even + rho_hat_odd) >= 0) {
+            rho_hat_t(t + 1) = rho_hat_even;
+            rho_hat_t(t + 2) = rho_hat_odd;
           }
-          rho_hat = 1 - (mean_var - mean(acov_t)) / var_plus;
-          if (rho_hat >= 0)
-            rho_hat_t(t) = rho_hat;
-          max_t = t;
+          max_t = t + 2;
         }
-        double ess = chains * n_samples;
-        if (max_t > 1) {
-          ess /= 1 + 2 * rho_hat_t.sum();
+        // Geyer's initial monotone sequence
+        for (int t = 3; t <= max_t - 2; t += 2) {
+          if (rho_hat_t(t + 1) + rho_hat_t(t + 2) >
+              rho_hat_t(t - 1) + rho_hat_t(t)) {
+            rho_hat_t(t + 1) = (rho_hat_t(t - 1) + rho_hat_t(t)) / 2;
```

Comment:
I'm pretty sure this logic is incorrect. I believe if `rho_hat_t(t + 1) + rho_hat_t(t + 2) > rho_hat_t(t - 1) + rho_hat_t(t)`, we want the `rho_hat_t` from `t + 1` until the end (or just out to `max_t`) to be 0.

I'm not sure why we're filling in the value with half of the sum of the larger value.

I think the right thing to do is to set `rho_hat_t(t + 1)` and `rho_hat_t(t + 2)` to 0 and we can leave the loop as-is.

Reply:
Also, should we start at `t = 1`? Is there any chance that the sum of the second pair is bigger than the first pair?

Reply:
The logic is correct. Geyer (1992) uses only the paired sums in description of monotone sequence. I can rewrite this to use pairs sums, if you think this is too confusing. If a paired sum is larger than the previous paired sum, we set it to the value of the previous paired sum. When we do this sequentially, the series of the paired sums is non-increasing. Since we care only about the paired sums, it really doesn't matter what are the actual values for rho's as long as the paired sum is correct. I just decided to divide the paired sum equally for two rho's in the sum. There is a very small probability that second pair is larger than the first pair, and even then the difference has to be small.

Reply:
I finally got it. Yes, the logic is right. It's not confusing at all.

I was having trouble picking it out of the wording: "The initial monotone sequence is obtained by further reducing $Pˆt′$ to the minimum of the preceding ones so that the estimated sequence is monotone."

I had misread the text, so if the original summed pair sequence was something like `(1.0, 0.9, 0.2, 0.8, 0.7, 0.0)`, I thought it should have been `(1.0, 0.9, 0.2, 0.0, 0.0, 0.0)`, but after re-reading, your code produces the right thing: `(1.0, 0.9, 0.2, 0.2, 0.0)`.